# Exploring Ebay Car Sales Data

We have data of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website

We are going to first clean the data for further investigation

Second we are going to find mean price and mean kilometer for top brands and compare them 

## Openeing the data set

First we open the csv file including the data set by using panda


In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding ='Latin-1' )

Hear by just running the data frame in Jupyter we can access to some part of data

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

And here is the info function to display information about our data frame

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As we can see the data set has 50000 rows and 20 columns. Values are int64 or object (mostly string) and some columns have null values. Also we can see the name of each columns in the info function result and the table above.

## Modifying columns name

For the first step of data cleaning we change the columns name from camelCase to snakecase(Python's preferred)
Here in the code blow we can display columns name


In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We change the names and store them in cleaned_columns and assign them back to autos columns

In [6]:
cleaned_columns = ['date_crawled', 'name', 'seller', 'type_offer', 'price', 'abtest',
       'type_vehicle', 'registration_year', 'gearbox', 'ps_power', 'model',
       'odometer', 'registration_month', 'type_fuel', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_nr', 'code_postal',
       'last_seen']

autos.columns = cleaned_columns

autos.columns

Index(['date_crawled', 'name', 'seller', 'type_offer', 'price', 'abtest',
       'type_vehicle', 'registration_year', 'gearbox', 'ps_power', 'model',
       'odometer', 'registration_month', 'type_fuel', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_nr', 'code_postal',
       'last_seen'],
      dtype='object')

## Exploring and cleaning the data

In this step we use describe function to explore the data.

As we can see the columns seller and type_offer have 2 unique values and freq is equal to 49999, it means there is only one value for this column and we can remove these columns from the data set

The price and the odometer columns are numerical columns but with some text in them and we will remove the strings from them and convert them to numerical value later.




In [7]:
autos.describe(include = 'all')

date_crawled         name  seller type_offer  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       type_vehicle  registration_year  gearbox      ps_power  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month type_fuel       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_nr   code_postal            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

pictures_nr column needs more investigation and the result of blow code shows us the whole column is actually is zero and we can remove this column too

In [8]:
# power_is_null = autos["registration_month"].isnull()

# print(power_is_null)

print(autos["pictures_nr"].value_counts())

print(autos["pictures_nr"].head())
autos["pictures_nr"].unique()

0    50000
Name: pictures_nr, dtype: int64
0    0
1    0
2    0
3    0
4    0
Name: pictures_nr, dtype: int64


array([0])

In this step we will do the modification which explained lately

We first start by removing three columns pictures_nr, seller and type_offer


In [9]:
autos =  autos.drop(['pictures_nr', 'seller', 'type_offer'], axis = 1)

In this step we will delete the strings in price and odometer columns and convert them to int64 type

In [10]:
autos['price'] = autos['price'].str.replace("$","")
autos['price'] = autos['price'].str.replace(",","").astype(int)

autos['odometer'] = autos['odometer'].str.replace("km","")
autos['odometer'] = autos['odometer'].str.replace(",","").astype(int)


Now we are going to change the name of price column to price_dollar and the odometer to odometer_km to better displaying the columns name. After that we showed some rows for these columns as you can see the changes.

In [11]:
autos.rename({'price':'price_dollar'}, axis = 1, inplace = True)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

print(autos['price_dollar'].head(5))
print(autos['odometer_km'].head(5))

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_dollar, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64


## Finding outliers in price_dollar and odometer_km

Here we about to investigate in price_dollar column to find outliers.

We first printed the number of unique values which is 2357. 
Then we printed the counted value in descending order to see highest and lowest prices with their counts. We can see there is car with almost 100 million dollar and 0 dollar. Also if we investigate the brand and the type of cars with price of more than 1 million dollar we can see the results in blow. I 'm sure we have not used fiat with more than 1 million dollar price. Therefore the cars with price of than 1 million is outliers.

Let's investigate lower price cars brand and type_vehicle and unrepaired_damage. We can see BMW Audi and etc. for less than 50 dollar witch I think they are outliers and we are going to remove all outliers.


In [11]:
print(autos['price_dollar'].unique().shape)

print(autos['price_dollar'].describe())

print(autos['price_dollar'].value_counts().sort_index(ascending = False))

# for more than 1 million price cars
miliondollars = autos.loc[autos["price_dollar"] > 1000000 , ["brand", "type_vehicle"]]

print(miliondollars)

# for les than 50 dollar cars
fifty_lower = autos.loc[autos["price_dollar"] < 50 , ["brand", "type_vehicle", "unrepaired_damage"]]

print(fifty_lower)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64
99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4

After finding outliners it is the time to remove them as in code blow. We only kept the rows with price higher than 50 and lower than 1 million. We assign new data set to auto_price_clean data frame. Also we displayed some information about our cleaned price_dollar column.

In [12]:
autos = autos[autos["price_dollar"].between(50, 1000000)]


print(autos['price_dollar'].describe())




count     48350.000000
mean       5977.136381
std       11983.401860
min          50.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      999999.000000
Name: price_dollar, dtype: float64


In this step we are going to find the outliners in odometer_km column. As it is clear from the result in blow the min value 5000 km and the max value is 15000 witch both of them is rational value for odometer and there is no outliner in this column


In [13]:
print(autos['odometer_km'].unique().shape)

print(autos['odometer_km'].describe())

print(autos['odometer_km'].value_counts().sort_index(ascending = False))


(13,)
count     48350.000000
mean     125876.732161
std       39617.575951
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    31290
125000     5046
100000     2105
90000      1733
80000      1414
70000      1215
60000      1153
50000      1011
40000       815
30000       778
20000       759
10000       249
5000        782
Name: odometer_km, dtype: int64


## Exploring and correcting date columns

In this step we are going to investigate in the columns that store dates witch include 
`date_crawled`
`last_seen`
`ad_created`
`registration_month`
`registration_year`

The first three of them are in string and the last two ones are numerical. For the strings one we can print some of them to see the format for each column as in code blow.


In [14]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

For all of them first 10 characters are for only the date. We can use this to extract only the dates and generate distribution for the dates in these three columns as in code blow

In [15]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna = False).sort_index())


2016-03-05    0.025357
2016-03-06    0.014064
2016-03-07    0.036029
2016-03-08    0.033257
2016-03-09    0.033051
2016-03-10    0.032223
2016-03-11    0.032554
2016-03-12    0.036980
2016-03-13    0.015657
2016-03-14    0.036670
2016-03-15    0.034292
2016-03-16    0.029493
2016-03-17    0.031603
2016-03-18    0.012906
2016-03-19    0.034747
2016-03-20    0.037787
2016-03-21    0.037311
2016-03-22    0.032947
2016-03-23    0.032265
2016-03-24    0.029431
2016-03-25    0.031499
2016-03-26    0.032265
2016-03-27    0.031086
2016-03-28    0.034891
2016-03-29    0.034126
2016-03-30    0.033713
2016-03-31    0.031830
2016-04-01    0.033733
2016-04-02    0.035533
2016-04-03    0.038573
2016-04-04    0.036525
2016-04-05    0.013051
2016-04-06    0.003164
2016-04-07    0.001386
Name: date_crawled, dtype: float64


the distribution for date _crawled is somehowe uniform during almost one month

In [16]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna = False).sort_index())


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033133
2016-03-10    0.031934
2016-03-11    0.032885
2016-03-12    0.036794
2016-03-13    0.017022
2016-03-14    0.035305
2016-03-15    0.034023
2016-03-16    0.029990
2016-03-17    0.031272
2016-03-18    0.013588
2016-03-19    0.033630
2016-03-20    0.037849
2016-03-21 

There is a lot of variety for the date that the is created

In [17]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna = False).sort_index())

2016-03-05    0.001075
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007342
2016-03-09    0.009638
2016-03-10    0.010610
2016-03-11    0.012389
2016-03-12    0.023744
2016-03-13    0.008873
2016-03-14    0.012637
2016-03-15    0.015884
2016-03-16    0.016463
2016-03-17    0.028087
2016-03-18    0.007322
2016-03-19    0.015801
2016-03-20    0.020620
2016-03-21    0.020600
2016-03-22    0.021406
2016-03-23    0.018594
2016-03-24    0.019772
2016-03-25    0.019173
2016-03-26    0.016774
2016-03-27    0.015574
2016-03-28    0.020869
2016-03-29    0.022337
2016-03-30    0.024736
2016-03-31    0.023826
2016-04-01    0.022875
2016-04-02    0.024881
2016-04-03    0.025171
2016-04-04    0.024509
2016-04-05    0.124964
2016-04-06    0.221634
2016-04-07    0.132079
Name: last_seen, dtype: float64


Last seen is the date When the crawler saw this ad last online. It is the date when the ad removed because the car was sold. There's are more cars that has been sold the last three days

In [18]:
autos['registration_year'].describe()

count    48350.000000
mean      2004.718035
std         87.783995
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There are something odd in registration_year as we can see the min year is 1000 and the max is 9999 witch is nonsense for the year and we need to deal with this. Also the car can't be registered after the listing was seen. Therefore any number 2016 is inaccurate


We need to remove any value for this column bigger than 2016 and lower than 1900.

With Boolean indexing we are going to keep rows with registration_year between 2016 and 1900.


In [19]:
autos = autos.loc[autos["registration_year"].between(1900,2016)]

In [20]:
autos['registration_year'].value_counts(normalize = True)

2000    0.067131
2005    0.062850
1999    0.062054
2004    0.058073
2003    0.057987
2006    0.057449
2001    0.056631
2002    0.053318
1998    0.050585
2007    0.048928
2008    0.047594
2009    0.044797
1997    0.041742
2011    0.034857
2010    0.034190
1996    0.029348
2012    0.028165
2016    0.025970
1995    0.025927
2013    0.017235
2014    0.014244
1994    0.013534
1993    0.009123
2015    0.008262
1992    0.007918
1991    0.007294
1990    0.007230
1989    0.003679
1988    0.002883
1985    0.002044
          ...   
1966    0.000473
1976    0.000452
1969    0.000409
1975    0.000387
1965    0.000366
1964    0.000258
1963    0.000172
1961    0.000129
1959    0.000129
1962    0.000086
1956    0.000086
1937    0.000086
1958    0.000086
1955    0.000043
1954    0.000043
1957    0.000043
1951    0.000043
1910    0.000043
1934    0.000043
1941    0.000043
1927    0.000022
1929    0.000022
1950    0.000022
1931    0.000022
1948    0.000022
1938    0.000022
1939    0.000022
1953    0.0000

Most of the cars registration date is about 20 years ago

## Finding common brands and their mileage

In this part we will find top brands and their distant traveled to see the relation


In [21]:
unique_brands = autos["brand"].unique()
print(unique_brands)

autos["brand"].value_counts(normalize = True)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


volkswagen        0.211356
bmw               0.109949
opel              0.107303
mercedes_benz     0.096501
audi              0.086604
ford              0.069993
renault           0.047186
peugeot           0.029908
fiat              0.025712
seat              0.018246
skoda             0.016417
nissan            0.015320
mazda             0.015255
smart             0.014179
citroen           0.014050
toyota            0.012759
hyundai           0.010048
sonstige_autos    0.009618
volvo             0.009101
mini              0.008779
mitsubishi        0.008176
honda             0.007854
kia               0.007100
alfa_romeo        0.006649
porsche           0.006046
suzuki            0.005939
chevrolet         0.005702
chrysler          0.003529
dacia             0.002647
daihatsu          0.002517
jeep              0.002281
subaru            0.002109
land_rover        0.002109
saab              0.001657
jaguar            0.001571
daewoo            0.001485
trabant           0.001377
r

As you cans see the first brand is Volkswagen with 21% then BMW with 10%. Actually top 5 brands are German which shows the domination by brands from this country. Also we choose top 10 brands to aggregate by.

In [22]:
top_10 = autos["brand"].value_counts(normalize = True)[:10].index

print(top_10)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [23]:
brands_mean_price = {}

for brand in top_10:
    mean_price = autos.loc[autos["brand"] == brand, 'price_dollar'].mean()
    brands_mean_price[brand] = mean_price
    
print(brands_mean_price)

{'volkswagen': 5627.416675150158, 'bmw': 8376.805088062622, 'opel': 2996.0850210547424, 'mercedes_benz': 8663.059308807135, 'audi': 9373.771428571428, 'ford': 4068.368275438057, 'renault': 2483.889192886457, 'peugeot': 3100.6928057553955, 'fiat': 2818.456066945607, 'seat': 4423.116745283019}


As we can see the results the high mean price is or Audi brand and then Mercedes Benz then BMW.

We are going to fine mean mileage for top 10 brands as we did for mean price. The code will be like the mean price section


In [24]:
brands_mean_milage = {}

for brand in top_10:
    mean_price = autos.loc[autos["brand"] == brand, 'odometer_km'].mean()
    brands_mean_milage[brand] = mean_price
    
print(brands_mean_milage)

{'volkswagen': 128775.32322101192, 'bmw': 132651.66340508807, 'opel': 129440.54541808703, 'mercedes_benz': 130923.07692307692, 'audi': 129254.65838509316, 'ford': 124296.03442975714, 'renault': 128144.09484724123, 'peugeot': 127122.30215827338, 'fiat': 117066.94560669456, 'seat': 121303.06603773584}


In this part we are going to first convert both dictionaries into series object using series constructor then create data frame using first series by data frame constructor. Assign the other one to the data frame then print the data frame

In [25]:
bmp_series = pd.Series(brands_mean_price)
bmm_series = pd.Series(brands_mean_milage)
df = pd.DataFrame(bmp_series, columns=['mean_price'])

df["mean_milage"] = bmm_series

print(df)

                mean_price    mean_milage
volkswagen     5627.416675  128775.323221
bmw            8376.805088  132651.663405
opel           2996.085021  129440.545418
mercedes_benz  8663.059309  130923.076923
audi           9373.771429  129254.658385
ford           4068.368275  124296.034430
renault        2483.889193  128144.094847
peugeot        3100.692806  127122.302158
fiat           2818.456067  117066.945607
seat           4423.116745  121303.066038


Very slightly the expensive cars have been drove more. But it is not very distinguishable.

## Extra cleaning

To clean the data set more we start first with translating German words to English.
Then we will convert string dates to numerical ones 
### German words to English

If we look at the data carefully we can realize in some columns there are German words and we can see the by the code blow by showing unique values for each that columns.


In [26]:
print(autos['type_vehicle'].unique())
print(autos['gearbox'].unique())
print(autos['type_fuel'].unique())
print(autos['unrepaired_damage'].unique()) 


['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


As we can see in above we shouel translate some words:

kleinwagen > small car

kombi > station wagon

andere > other

benzin > petrol

elektro > electrical

automatik > automatic

nein > no

ja > yes


in the code blow we replay the german words we their english counterparts

In [27]:
autos['type_vehicle'] = autos['type_vehicle'].str.replace("kleinwagen", "small car")
autos['type_vehicle'] = autos['type_vehicle'].str.replace("kombi", "station wagon")

autos['gearbox'] = autos['gearbox'].str.replace("automatik", "automatic")

autos['type_fuel'] = autos['type_fuel'].str.replace("benzin", "petrol")
autos['type_fuel'] = autos['type_fuel'].str.replace("elektro", "other")
autos['type_fuel'] = autos['type_fuel'].str.replace("andere", "other")

autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("nein", "no")
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("ja", "yes")

autos['model'] = autos['model'].str.replace("andere", "other")

### String dates to numerical values

In this section we will convert dates to numerical values for example

2016-05-25 to 20160525 

We first replace the '-' with empty character then we will remove the time after that we will convert the type to integer 
We did this changes to columns date_crawled, ad_created  and last_seen  as in code blow


In [28]:
autos['date_crawled'] = autos['date_crawled'].str.replace("-", "")
autos['date_crawled'] = autos['date_crawled'].str[0:8].astype(int)

autos['ad_created'] = autos['ad_created'].str.replace("-", "")
autos['ad_created'] = autos['ad_created'].str[0:8].astype(int)

autos['last_seen'] = autos['last_seen'].str.replace("-", "")
autos['last_seen'] = autos['last_seen'].str[0:8].astype(int)

In [29]:
autos

date_crawled                                               name  \
0          20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1          20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2          20160326                         Volkswagen_Golf_1.6_United   
3          20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4          20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5          20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6          20160320  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7          20160316                               Golf_IV_1.9_TDI_90PS   
8          20160322                                         Seat_Arosa   
9          20160316          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11         20160316                         Mercedes_A140_Motorschaden   
12         20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13         20160323                                 Audi_A3_1.6_tuning   
14         20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15         20160401        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16         20160316                                Opel_Vectra_B_Kombi   
17         20160329                          Volkswagen_Scirocco_2_G60   
18         20160326               Verkaufen_mein_bmw_e36_320_i_touring   
19         20160317        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20         20160305  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21         20160306                      Porsche_911_Carrera_4S_Cabrio   
22         20160328                               MINI_Cooper_S_Cabrio   
23         20160310   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24         20160403                         BMW_535i_xDrive_Sport_Aut.   
25         20160321  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26         20160403                                Volkswagen_Polo_Fox   
28         20160319                                      MINI_Cooper_D   
29         20160402  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30         20160314                         Peugeot_206_Unfallfahrzeug   
31         20160314                          Noch_gut_erhaltenen_C_320   
...             ...                                                ...   
49968      20160401                  Mercedes_Benz_190_D_2.5_Automatik   
49969      20160317             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970      20160321  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971      20160329                                         W.Lupo_1.0   
49972      20160326          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973      20160327                   Mercedes_Benz_SLK_200_Kompressor   
49975      20160327          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976      20160319  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977      20160331                        Mercedes_Benz_C200_Cdi_W203   
49978      20160404                        Mercedes_Benz_E_200_Classic   
49979      20160320                      Volkswagen_Polo_1.6_TDI_Style   
49980      20160312                            Ford_Escort_Turnier_16V   
49981      20160315             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982      20160329            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983      20160306                                      Ford_focus_99   
49985      20160402                          Verkaufe_meinen_vw_vento!   
49986      20160404  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987      20160322  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...   
49988      20160328                                         BMW_330_Ci   
49989      20160311        VW_Polo_zum_Ausschlachten_oder_Wiederaufbau   
49990      20160321         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban   
49991      20160306                                         Kleinwagen   
49992      

## finding most common brand/model combination 

In this step we first define empty dictionary
Then fin top 10 model
After that we find corresponding model brand and the model's frequency

For dictionary key we use brand + space + model name and for the value the size of brand_model sries witch is the number of that model. Then for better view we construct a data frame from brand_model_top.

Note that we remove the 'other' model from top models


In [89]:
brand_model_top = {}


Top_10_model = autos["model"].value_counts(ascending = False)[:10]

#dropping other lable
Top_10_model = Top_10_model.drop(labels="other")

for m in Top_10_model.index:
    brand_model = autos.loc[autos["model"] == m, 'brand']
    brand = brand_model.iloc[0]
    brand_model_top[brand + ' ' + m] = brand_model.size
    



brand_model_series = pd.Series(brand_model_top)

df2 = pd.DataFrame(brand_model_series, columns=['counts'])
df2

As we can see the most common brand model is Volkswagen golf then BMW 3er and etc.

## Conclusion 

We first cleaned the data then we investigated the brand prices and concludes that Audi brand and then Mercedes Benz then BMW are the most expensive ones. Also there was a slight relation between the cars price and odometer that expensive cars have more mileage. Also we sow the most common brand models are Volkswagen golf then BMW 3er and etc.
